88.md

```diff
@@ -0,0 +1,107 @@
+    Id, Pubkey, Timestamp, Kind, Tags, Content,
+    -- Type-length-value encoded stream:
+    TLVs,
+    -- NIP-1 signature
+    legacy Signature
+    -- new signature over binary data (except legacy sig)
+    binary Signature
+
+Id :: [Byte ^ 32]     -- indicates fixed-size byte string of exactly 32 bytes
+Pubkey :: [Byte ^ 32]
+Timestamp :: U32 -- indicates 32-bit unsigned integer
+Kind :: U64 -- indicates 64-bit unsigned integer
+
+Tags :: [Tag ^ 0..0xFFF] -- indicates array of tags with up to 4096 elements
+Tag :: TagName, TagData
+TagName :: [AsciiPrintable ^ 1..0xFF] -- indicates string from 1 to 255 chars
+                                      -- containing only printable ASCII chars
+                                      -- (bytes 0x20-0x7e)
+TagData :: [Entry ^ 0..0xFF]
+Entry :: [Byte ^ 1..0xFF]     -- inticates non-empty byte string up to 255 bytes
+
+Content :: [Byte ^ 0..0xFFFFFF] -- inticates byte string up to 2^24 bytes
+Signature :: [Byte ^ 64] -- indicates Shnorr signature
+
+-- Type-length-value encoded stream
+TLVs :: [TLVEntry ^ 0..0xFF]
+TLV :: type U64, value [Byte ^ 0..0xFFFF]
+```
+
+The pseudocode represents in-memory and network serialization layout of the 
+binary data, organized in form of fields listed after `::`. All integers are 
+serialized in little-endian format (to match the bitcoin protocol); hashes, 
+public keys and signatures match bitcoin consensus rules for serialization. 
+Variable-size collections (i.e. excluding id, pubkey and signatures) are 
+prefixed with either a single byte, a 16-bit or 24-bit word (in little-endian 
+encoding) matching the bit dimensionality for the maximum number of elements and 
+specifying the number of elements in the collection. The used notation follows
+[strict type] system notation for a memory binary data layout, which can't be
+deterministically specified if a protocol buffers or similar notation is used.
+
+The binary events are provided with two signatures: a legacy one, created over
+JSON-serialized message without TLV stream (see [Backward Compatibility](#backward-compatibility)
+section below); and a new signature, created over binary-serialized event data,
+including TLV stream but excluding the legacy signature.
+
+The maximal total length of the binary-serialized tags must not exceed 16MB 
+(0xFFFFFF bytes); thus the total length of the event can't be larger than 
+50 268 360 bytes (~50 MB).
+
+
+Backward compatibility
+----------------------
+
+Clients and relays not supporting the current proposal must ignore the events.
+
+Clients and relays supporting the proposal must provide a NIP-1-compatible 
+encoding of the binary data for the legacy (pre-NIP-88) clients and relay,
+where:
+* `content` is given in the Base64 encoding;
+* `signature` over binary data is skipped;
+* `tlv` stream is skipped;
+* `tag` entries are provided in lowercase hexadecimal (base64) encoding.
+
+Binary protocol should operate using different network protocol and port number,
```

Comment:
Couldn't this binary protocol be delivered over existing web-socket, either on a different URL's path, or when certain http header is present?

Reply:
It could be over the same websocket just sent as a bytes message instead of a text message

Reply:
It would make sense if a relay/client decide on binary support, that they both start using the binary message protocol.